tools/vhost-user-video/vhost-user-video.c

```diff
@@ -0,0 +1,1824 @@
+ */
+
+#define G_LOG_DOMAIN "vhost-user-video"
+#define G_LOG_USE_STRUCTURED 1
+
+#include <glib.h>
+#include <gio/gio.h>
+#include <gio/gunixsocketaddress.h>
+#include <glib-unix.h>
+#include <glib/gstdio.h>
+#include <stdio.h>
+#include <string.h>
+#include <inttypes.h>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <endian.h>
+#include <assert.h>
+
+#include "libvhost-user-glib.h"
+#include "libvhost-user.h"
+#include "standard-headers/linux/virtio_video.h"
+
+#include "qemu/compiler.h"
+#include "qemu/iov.h"
+
+#include "vuvideo.h"
+#include "v4l2_backend.h"
+#include "virtio_video_helpers.h"
+
+#ifndef container_of
+#define container_of(ptr, type, member) ({                      \
+        const typeof(((type *) 0)->member) * __mptr = (ptr);     \
+        (type *) ((char *) __mptr - offsetof(type, member)); })
+#endif
+
+static gchar *socket_path;
+static gchar *v4l2_path;
+static gint socket_fd = -1;
+static gboolean print_cap;
+static gboolean verbose;
+static gboolean debug;
+
+static GOptionEntry options[] = {
+    { "socket-path", 0, 0, G_OPTION_ARG_FILENAME, &socket_path,
+      "Location of vhost-user Unix domain socket, "
+      "incompatible with --fd", "PATH" },
+    { "v4l2-device", 0, 0, G_OPTION_ARG_FILENAME, &v4l2_path,
+      "Location of v4l2 device node", "PATH" },
+    { "fd", 0, 0, G_OPTION_ARG_INT, &socket_fd,
+      "Specify the fd of the backend, "
+      "incompatible with --socket-path", "FD" },
+    { "print-capabilities", 0, 0, G_OPTION_ARG_NONE, &print_cap,
+      "Output to stdout the backend capabilities "
+      "in JSON format and exit", NULL},
+    { "verbose", 'v', 0, G_OPTION_ARG_NONE, &verbose,
+      "Be more verbose in output", NULL},
+    { "debug", 0, 0, G_OPTION_ARG_NONE, &debug,
+      "Include debug output", NULL},
+    { NULL }
+};
+
+enum {
+    VHOST_USER_VIDEO_MAX_QUEUES = 2,
+};
+
+static const char *
+vv_cmd_to_string(int cmd)
+{
+#define CMD(cmd) [cmd] = #cmd
+    static const char *vg_cmd_str[] = {
+        /* Command */
+        CMD(VIRTIO_VIDEO_CMD_QUERY_CAPABILITY),
+        CMD(VIRTIO_VIDEO_CMD_STREAM_CREATE),
+        CMD(VIRTIO_VIDEO_CMD_STREAM_DESTROY),
+        CMD(VIRTIO_VIDEO_CMD_STREAM_DRAIN),
+        CMD(VIRTIO_VIDEO_CMD_RESOURCE_CREATE),
+        CMD(VIRTIO_VIDEO_CMD_RESOURCE_QUEUE),
+        CMD(VIRTIO_VIDEO_CMD_RESOURCE_DESTROY_ALL),
+        CMD(VIRTIO_VIDEO_CMD_QUEUE_CLEAR),
+        CMD(VIRTIO_VIDEO_CMD_QUERY_CONTROL),
+        CMD(VIRTIO_VIDEO_CMD_GET_CONTROL),
+        CMD(VIRTIO_VIDEO_CMD_SET_CONTROL),
+        CMD(VIRTIO_VIDEO_CMD_GET_PARAMS_EXT),
+        CMD(VIRTIO_VIDEO_CMD_SET_PARAMS_EXT),
+    };
+#undef CMD
+
+    if (cmd >= 0 && cmd < G_N_ELEMENTS(vg_cmd_str)) {
+        return vg_cmd_str[cmd];
+    } else {
+        return "unknown";
+    }
+}
+
+/* taken from util/iov.c */
+size_t video_iov_size(const struct iovec *iov, const unsigned int iov_cnt)
```

Comment:
If this function is only used here it can be static.

Reply:
Actually as the comment suggest, this (and the others to_buffer and from_buffer) function is a direct copy of the code at `iov.c`. So I think I will remove them and use the imported version instead to avoid duplication.